Created BoundingBox class, road_node bounding box and road bounding box

jazminlaila opened the pull request:
I created the BoundingBox class at bounding_box.py and it's tests. I added a bounding box method to RoadNode class and also it's tests. I also added a method (which I'm not sure if ti is OK) to Road class. The tests for this last method are commented because they don't work yet, but may work in the future.

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,33 @@
+class BoundingBox(object):
+
+    def _normalize(self):
+        new_origin = Point(min(self.origin.x, self.corner.x),
```
This is a typical case where it would be handy to just do `self.origin.min(self.corner)`, not only because then you wouldn't have to repeat a very similar code in `merge`, but because you could also write the test for it the `Point` test class and then be more confident of your code. The same applies for `max`

jazminlaila replied:
You want the max between the two points giving points an order, or a point like the one I wrote? (Yes, it's not the same, it depends on which opposite corners we give)

basicNew replied:
Since ordering points in 3D (and 2D for that matter) is tricky, I would say we define the min of two points as the min of each component (same for max)

jazminlaila replied:
so the min of tho points may be neither of both points... weird, but ok

jazminlaila replied:
I really don't like calling it 'min'

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,33 @@
+
+class BoundingBox(object):
+
+    def _normalize(self):
```
Could you move the private method at the end of the class definition?

jazminlaila replied:
done

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,33 @@
+        self.corner = corner
+        self._normalize()
+
+    def __eq__(self, other):
```
+1 for redefining the equality! Please ping me when you have a moment so we can briefly discuss the `__hash__` implementation you must also add when redefining the equality.

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,33 @@
+                             max(self.corner.y, other.corner.y))
+        return BoundingBox(merge_origin, merge_corner)
+
+    def merge_multiple_boundingboxes(self, box_list):
```
While the idea is ok, it seems weird that you are asking a bounding box to merge other bounding boxes but not to participate at all in the process or final result. Like you can do:

```
boxes = [BoundingBox(Point(4, 4), Point(5, 5)),
              BoundingBox(Point(4.5, 4.5), Point(6, 6))]
 BoundingBox(Point(0, 0), Point(1, 1)).merge_multiple_boundingboxes(boxes)
```
and the resulting bounding box has nothing to do with `BoundingBox(Point(0, 0), Point(1, 1))`

What you want instead is to be able to create a new bounding box from a collection of bounding boxes. To do that you need as a class method that acts as an alternative constructor so you can do things like:

```
boxes = [BoundingBox(Point(4, 4), Point(5, 5)),
              BoundingBox(Point(4.5, 4.5), Point(6, 6))]
BoundingBox.from_boxes(boxes)
```

the implementation would be quite similar to yours:

```
@classmethod
def from_boxes(cls, box_list):
       whole_merge = box_list[0]
        for i in range(1, len(box_list)):
            whole_merge = whole_merge.merge(box_list[i])
        return whole_merge
```

jazminlaila replied:
nice, I didn't like the way I did it, but didn't know other way to do it. That's looks muuuuuuuuch better. Thanks

basicNew commented on terminus/models/road_node.py:
```diff
@@ -1,11 +1,17 @@
         self.center = center
         self.name = name
 
+    def road_node_bounding_box(self, width):
```
just `bounding_box(self, width)` should be ok as a name

basicNew commented on terminus/models/road.py:
```diff
@@ -130,6 +132,17 @@ def _add_node(self, node):
+            node_bounding_boxes.append(node.road_node_bounding_box(self.width()))
+        return node_bounding_boxes
+
+    def road_bounding_box(self):
```
just `bounding_box(self)` is ok as a name

basicNew commented on terminus/models/road.py:
```diff
@@ -130,6 +132,17 @@ def _add_node(self, node):
+        return node_bounding_boxes
+
+    def road_bounding_box(self):
+        node_bounding_boxes = self.node_bounding_boxes
```
Missing a parenthesis here? `self.node_bounding_boxes()`

jazminlaila replied:
yes

basicNew commented on terminus/models/road.py:
```diff
@@ -130,6 +132,17 @@ def _add_node(self, node):
+        node_bounding_boxes = self.node_bounding_boxes
+        base_box = BoundingBox(Point(0, 0), Point(10, 10))
+        return base_box.merge_multiple_bounding_boxes(node_bounding_boxes)
+
```
If you apply what I mentioned earlier this method would just become: 

```
return BoundingBox.from_boxes(self.node_bounding_boxes())
```

jazminlaila replied:
done

jazminlaila commented:
@basicNew I've updated this following the notes of your review and added a hash.

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,35 @@
+        return BoundingBox(merge_origin, merge_corner)
+
+    @classmethod
+    def from_boxes(cls, box_list):
```
Very minor: can we put this method right below `__init__`?

jazminlaila replied:
Done

basicNew commented on terminus/models/road.py:
```diff
@@ -130,6 +132,13 @@ def _add_node(self, node):
         self._nodes.append(node)
         node.added_to(self)
 
+    def node_bounding_boxes(self):
```
Shouldn't this be a private method?

jazminlaila replied:
ok

basicNew commented on terminus/tests/bounding_box_test.py:
```diff
@@ -0,0 +1,72 @@
+        expected_merge = BoundingBox(Point(10, -10), Point(50, 30))
+        self.assertEqual(box_1.merge(box_2), expected_merge)
+
+    def test_merge_nested_boxes(self):
```
Maybe you could add a test merging a bb with itself.

jazminlaila replied:
done

basicNew commented on terminus/tests/bounding_box_test.py:
```diff
@@ -0,0 +1,72 @@
+        box_1 = BoundingBox(Point(-5, -10), Point(5, 0))
+        box_2 = BoundingBox(Point(-5, 10), Point(5, 20))
+        box_list = [box_1, box_2]
+        expected_merge = box_1.merge(box_2)
```
Ideally you shouldn't rely on the `merge` method here, but create the expected bounding box "by hand", with the expected values.

jazminlaila replied:
I thought that as I wasn't testing the merge method here, it was 'legal' to use it (specially when what I wanted to check was that merging a list of two boxes was the same as using the merge method).

basicNew replied:
While it is legal to use it there are a couple of reasons not to:

- It becomes clearer to the reader what to expect. Otherwise in the code you don't get to see the values of the expected result.
- There are other ways to check that `from_boxes` uses that method (more advanced topics on testing). However in this case it would be nice to test this as a black box, not knowing how it performs the merge, so you can state your test as "If I give you this and that you should give this other thing back"

jazminlaila replied:
OK. Done

basicNew commented on terminus/tests/bounding_box_test.py:
```diff
@@ -0,0 +1,72 @@
+        expected_merge = BoundingBox(Point(-20, -30), Point(20, 20))
+        self.assertEqual(box_1.merge(box_2), expected_merge)
+
+    def test_merge_multiple_boundig_boxes_for_two_boxes(self):
```
Note that you are not testing the `merge` method here, but the `from_boxes` one. You should change the names of the test methods to reflect this.

jazminlaila replied:
Oh, it was for the old 'merge_multiple_bounding_boxes' method, that was changed into the class method from_boxes. I forgot to change the tests names :s

basicNew commented on terminus/tests/bounding_box_test.py:
```diff
@@ -0,0 +1,72 @@
+        expected_merge = BoundingBox(Point(-20, -10), Point(30, 20))
+        self.assertEqual(BoundingBox.from_boxes(box_list), expected_merge)
+
+    def test_merge_multiple_boundingboxes_for_five_boxes(self):
```
For the `from_boxes` these should be the interesting cases:

- Empty collection.
- Collection with 1 element.
- Collection with 2 elements.
- Collection with more than 2 elements.

jazminlaila replied:
except for the empty collection, those tests were already written, but with the wrong name...

basicNew replied:
Maybe I'm wrong, but IIRC there were tests for 2, 3 and 5 elements and what I'm saying is that there should be tests for 0, 1, 2 and +2 elements (again, maybe I overlooked something, just want to stress these are edge cases. I don't see the difference between testing 3 and 5 elements).

jazminlaila replied:
We are not supposed to obtain a box from an empty collection. Am I supposed to write an Error message for this case?

basicNew replied:
Exactly! :) . You can google a bit on exceptions in general and in Python in particular. Also take a look at https://en.wikibooks.org/wiki/Python_Programming/Errors and https://en.wikibooks.org/wiki/Python_Programming/Exceptions for a quick overview and https://jeffknupp.com/blog/2013/02/06/write-cleaner-python-use-exceptions/ for a more detailed article. Also see http://stackoverflow.com/questions/253314/exceptions-or-error-codes and http://stackoverflow.com/questions/99683/which-and-why-do-you-prefer-exceptions-or-return-codes for some discussion on exceptions vs return codes.

And returning to this particular case, you just just throw a `ValueError` when the collection is empty and write the test case for that (see e.g. http://stackoverflow.com/questions/129507/how-do-you-test-that-a-python-function-throws-an-exception)

jazminlaila replied:
Oh, you are right, I forgot the one element list.
Nice, I'll write the exception then (once I've learned how to).

jazminlaila replied:
It didn't work with ValueError. I don't understand much about it, but it seems it is not compatible with 'IndexError: list index out of range'. I tried something that passes the test, but I don't know if it is correct.

basicNew replied:
Hard to say without seeing the code. Let me know when you push the update and I'll take a look.

jazminlaila replied:
ok

basicNew commented on terminus/tests/road_node_test.py:
```diff
@@ -0,0 +1,15 @@
+
+class RoadNodeTest(unittest.TestCase):
+
+    def test_road_node_bounding_box(self):
```
Another case to test: what happens if `width == 0`?

jazminlaila replied:
Probably a box with same origin and corner, I think there shouldn't be a problem with this. Ok, I'll write the test.

basicNew replied:
Also: what happens if the width is negative?

jazminlaila replied:
Why would that happen? I suppose it would return the same box as the one with its absolute value. I can write an error message if necessary, but I think that is just not supposed to happen.

basicNew replied:
> Why would that happen?
That is a fair question and in the most common scenario because of a bug in the code calling it. However the aim of my question is more to think about all that comes together with defining a method, what to test and what not, etc. Couple of things to consider:

- Going back to "Why would that happen?" I may also say, "because it can happen". So the fact that it doesn't make sense, doesn't mean it can't happen. That is an important thing to always keep in mind.
- "I suppose it would return..." that is what tests are for, so that you (and later the user of your class) doesn't have to suppose, but get a well defined behavior. And it is important to mention that sometimes there is not a single answer to this; you could throw an error, take the absolute value of the width, decide to do a `max(0, width)` inside your implementation or let the bounding box be created with those values (which IIRC, due to the normalization, it will flip the origin and corner). The important thing is that you get a well defined behavior and your tests cover all the cases, even the "wrong" ones.

jazminlaila replied:
Ok. I chose to raise a ValueError,.

basicNew commented on terminus/tests/road_test.py:
```diff
@@ -66,3 +68,22 @@ def test_length_empty(self):
         self.assertEqual(road.length(), 0.0)
+
+        '''
+        this test couldn't be implemented yet:
```
Shouldn't you be able to test this now?

jazminlaila replied:
This is what happens:
ERROR: test_node_bounding_boxes (road_test.RoadTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jazmin/terminus/terminus/tests/road_test.py", line 79, in test_node_bounding_boxes
    road._add_node(node_1)
  File "/home/jazmin/terminus/terminus/models/road.py", line 133, in _add_node
    node.added_to(self)
  File "/home/jazmin/terminus/terminus/models/road_node.py", line 23, in added_to
    raise NotImplementedError()
NotImplementedError


is this 'NotImplementedError' something to be changed by some code in the future?
I wanted to create a road to test the method, but my problem was creating the road.

basicNew replied:
That is because you are creating an instance of `RoadNode`, which is an abstract class. You should create either a `RoadIntersectionNode` or a `RoadSimpleNode` for the test to work. Maybe it would be a good time to review a bit on OOP and python (e.g. https://python.swaroopch.com/oop.html and http://anandology.com/python-practice-book/object_oriented_programming.html) and we can later talk a bit about abstract classes.

jazminlaila replied:
I created the rod with 'from_points' method and it seemed to work :)

jazminlaila commented:
@basicNew , I've pushed the changes made after your last review. I still have to create the bounding boxes for the other classes.

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,38 @@
+                whole_merge = whole_merge.merge(box_list[i])
+            return whole_merge
+        except:
+            raise Exception('box_list should be a non empty list of bounding boxes')
```
Although this is a very pythonic way of implementing it, there is a small issue: the `try/except` block will catch any exception, not just the case that the `box_list` is empty. Even though there are better ways to implement this (we can discuss them later) I would suggest doing something along the lines of:

```
if not box_list:
    raise ValueError("box_list can't be empty")
whole_merge = box_list[0]
for i in range(1, len(box_list)):
    whole_merge = whole_merge.merge(box_list[i])
￼return whole_merge
```

basicNew commented on terminus/models/road.py:
```diff
@@ -130,6 +132,13 @@ def _add_node(self, node):
         self._nodes.append(node)
         node.added_to(self)
 
+    def __node_bounding_boxes(self):
```
Private methods use a single underscore. Also, can you move it to the bottom of the class definition?

basicNew commented on terminus/models/road_node.py:
```diff
@@ -1,11 +1,20 @@
+            return BoundingBox(box_origin, box_corner)
+        else:
+            raise ValueError('width should be a non negative number')
+
```
This is pretty much a matter of coding style, but we usually do this as:

```
if width < 0:
    raise ValueError('width should be a non negative number')
box_origin = Point(self.center.x - width / 2, self.center.y - width / 2)
box_corner = Point(self.center.x + width / 2, self.center.y + width / 2)
return BoundingBox(box_origin, box_corner)
```

basicNew commented on terminus/tests/bounding_box_test.py:
```diff
@@ -0,0 +1,84 @@
+
+    def test_from_boxes_with_one_box(self):
+        box_list = [BoundingBox(Point(17, 4), Point(59, 7))]
+        self.assertEqual(BoundingBox.from_boxes(box_list), BoundingBox(Point(17, 4), Point(59, 7)))
```
Minor thing, to avoid repeating the creation of the bb:

```
box = BoundingBox(Point(17, 4), Point(59, 7))
self.assertEqual(BoundingBox.from_boxes([box]), box))
```

basicNew commented on terminus/tests/bounding_box_test.py:
```diff
@@ -0,0 +1,84 @@
+        expected_merge = BoundingBox(Point(-20, -10), Point(30, 20))
+        self.assertEqual(BoundingBox.from_boxes(box_list), expected_merge)
+
+    def test_from_boxes_with_five_boxes(self):
```
I still don't see the difference between testing it with 3 or 5 boxes. Unless there is a good reason for it one of those two tests should be removed.

jazminlaila replied:
Just to be sure the first one hadn't passed casually?

basicNew replied:
Then why not adding another for 7? Or 10? Or 100? Please, don't get me wrong, not trying to make you angry, just trying to analyze what we test and how we test it. Ideally, each tests case should have a well defined purpose (or case to test); if for some reason you don't trust the test of 3 elements, then improve it so you can trust it, but having a big test codebase "just in case" doesn't scale (I know we barely talked about this, but one thing to keep in mind is that tests are also code; so you not only need to maintain / evolve the codebase of your program, but also the codebase of your tests. So having more tests means more work. This of course doesn't mean that you should write few tests, but that you should write tests for particular cases that you have identified and have a clear aim of what you are checking).

basicNew commented on terminus/tests/road_test.py:
```diff
@@ -66,3 +68,13 @@ def test_length_empty(self):
         road = Street.from_points([])
         self.assertEqual(road.length(), 0.0)
+
+    def test_node_bounding_boxes(self):
```
It would be nice to think the different cases that need testing. I would suggest starting with a single lane with offset 0, then a single lane with offset not 0 and finally one with multiple lanes (having positive and negative offsets).

jazminlaila commented:
@basicNew here's the last update. I've added a bounding box to the Building and Block classes and their tests (almost the same). I've also modified the min and max methods of the class Point into class methods. I still have to create a bounding box for the ground plane.

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,41 @@
+    @classmethod
+    def from_boxes(cls, box_list):
+        if not box_list:
+            raise ValueError("Box_list con't be empty")
```
knit: `box_list` and _can't_

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,41 @@
+        merge_corner = Point(max(self.corner.x, other.corner.x),
+                             max(self.corner.y, other.corner.y))
+        """
+        merge_origin = Point.min([self.origin, other.origin])
```
`min`/`max` shouldn't be class methods, but instance ones, so you can do:

```
merge_origin = self.origin.min(other.origin)
```

Also please remove the comment.

basicNew commented on terminus/geometry/bounding_box.py:
```diff
@@ -0,0 +1,41 @@
+        return BoundingBox(merge_origin, merge_corner)
+
+    def _normalize(self):
+        new_origin = Point.min([self.origin, self.corner])
```
Same as above

basicNew commented on terminus/geometry/point.py:
```diff
@@ -23,6 +23,20 @@ def from_tuple(cls, tuple):
             return cls(tuple[0], tuple[1])
 
+    @classmethod
+    def min(cls, points_list):
```
As per the comment above, you can move this to the instance side and have it take a single argument

basicNew commented on terminus/models/block.py:
```diff
@@ -30,3 +31,8 @@ def square(origin, size):
         generator.end_block(self)
+
+    def bounding_box(self):
+        box_origin = self.origin + Point.min(self.vertices)
```
Here you can apply the same method you used before: collect a set of bounding boxes (derived from the vertices) and do a `BoundingBox.from_boxes` to merge them all.

jazminlaila commented:
@basicNew , here there are the bounding boxes for ground plane and city and it's tests. I've made all of them 3D.

basicNew commented on terminus/models/city.py:
```diff
@@ -44,6 +49,22 @@ def add_intersection_at(self, point):
                     road.replace_node_at(point, intersection)
 
+    def bounding_box(self):
+        road_boxes = self._box_list_from_list(self.roads)
```
Although it is a slight variation, you can do `self._bounding_box_from_list(self.roads)`, where:

```
def _box_list_from_list(self, list):
    return BoundingBox.from_boxes(map(lambda x: x.bounding_box(), list))
```

this will save you a couple of lines of code.

basicNew commented on terminus/models/city.py:
```diff
@@ -55,3 +76,6 @@ def accept(self, generator):
             building.accept(generator)
         generator.end_city(self)
+
+    def _box_list_from_list(self, lst):
```
Remember not to use abbreviations; lst -> list

basicNew commented on terminus/models/ground_plane.py:
```diff
@@ -7,6 +9,12 @@ def __init__(self, size, origin, name=None):
         self.origin = origin
 
+    def bounding_box(self):
+        box_origin = Point(-self.size / 2, -self.size / 2)
```
Beware that you are doing integer division here (see e.g. http://stackoverflow.com/questions/2958684/python-division). You should do something along the lines of `x / 2.0`.

basicNew commented on terminus/tests/city_test.py:
```diff
@@ -57,3 +62,29 @@ def test_add_intersection_at_on_different_z_order(self):
+    def test_bounding_box_with_ground_plane(self):
+        building = Building.square(Point(35, 45), 10, 10)
+        block = Block.square(Point(-75, -45), 10)
+        road = Road.from_points([Point(60, -20), Point(80, -20)])
```
Create a street instead of creating a Road + adding a lane. A street is a road that has a single lane at offset 0.

jazminlaila commented:
@basicNew , I've pushed the last changes made following your last review.

basicNew commented on terminus/models/city.py:
```diff
@@ -44,6 +49,19 @@ def add_intersection_at(self, point):
+                                           groung_plane_bounding_box])
+        else:
+            return BoundingBox.from_boxes([roads_bounding_box, blocks_bounding_box,
+                                           buildings_bounding_box])
```
Sorry I missed this on the previous review. You can make this code smaller / easier to read by changing it a bit (writing this form the top of my head):

```
child_bounding_boxes = [self._bounding_box_from_list(self.roads),
                        self._bounding_box_from_list(self.blocks), 
                        self._bounding_box_from_list(self.buildings)]
if self.ground_plane is not None:
    child_bounding_boxes.append(self.ground_plane.bounding_box())
    
return BoundingBox.from_boxes(child_bounding_boxes)
```

jazminlaila commented:
@basicNew , I've pushed the last change you asked for.
